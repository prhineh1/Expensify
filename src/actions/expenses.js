import uuid from 'uuid';
import database from '../firebase/firebase';

//ADD_EXPENSE

export const addExpense = (expense = {}) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const addExpenseComplete = (expense = {}) => ({
    type: 'ADD_EXPENSE_COMPLETE',
    expense
});

//REMOVE_EXPENSE

export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

//EDIT_EXPENSE

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

//SET_EXPENSES

export const setExpenses = () => ({ type: 'SET_EXPENSES' });

export const setExpensesComplete = (expenses) => ({
    type: 'SET_EXPENSES_COMPLETE',
    expenses
});