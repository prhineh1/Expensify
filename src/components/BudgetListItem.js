import React from 'react';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import moment from 'moment';

const BudgetListItem = (props) => {
    const expensesTotal = props.expenses.reduce((sum, { amount }) => sum + amount, 0);
    const listItemClass = expensesTotal > props.budget.amount ? 'list-item list-item--over' : 'list-item';
    return ( 
        <Link className={listItemClass} to={'/edit/budget/'+ props.budget.id}>
            <div>
                <h3 className='list-item__title'>{props.budget.description}</h3>
            </div>
            <h4 className='list-item__budget'>
                <span className='list-item__budget__mobile'>Expenses: {numeral(expensesTotal / 100).format('$0,0.00')}</span>
                <span className='list-item__budget__desktop'>{numeral(expensesTotal / 100).format('$0,0.00')}</span>
            </h4>
            <h3 className='list-item__data'>
                <span className='list-item__data__mobile'>Budget: {numeral(props.budget.amount / 100).format('$0,0.00')}</span>
                <span className='list-item__data__desktop'>{numeral(props.budget.amount / 100).format('$0,0.00')}</span>
            </h3>
        </Link>
    )
};

export default BudgetListItem