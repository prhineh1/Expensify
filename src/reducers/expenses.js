//Expenses Reducer

const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
    switch(action.type) {
        case 'ADD_EXPENSE_COMPLETE':
            return [...state, action.expense];
        case 'REMOVE_EXPENSE_COMPLETE':
            return state.filter(({id}) => id !== action.id);
        case 'EDIT_EXPENSE_COMPLETE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    };
                } else {
                    return expense;
                }
            });
        case 'SET_EXPENSES_COMPLETE':
            return action.expenses;
        default:
            return state;
    }
};
