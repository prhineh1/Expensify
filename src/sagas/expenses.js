import ReactDOM from 'react-dom';
import { takeEvery } from 'redux-saga';
import { put, all, call } from 'redux-saga/effects'
import * as Api from '../firebase/firebase';

//WATCHERS

export function* watchAddExpenseSaga() {
    yield takeEvery('ADD_EXPENSE', addExpenseAsync);
};

export function* watchSetExpensesSaga() {
    yield takeEvery('SET_EXPENSES', setExpensesAsync);
};

export function* watchRemoveExpenseSaga() {
    yield takeEvery('REMOVE_EXPENSE', removeExpenseAsync);
};

export function* watchEditExpenseSaga() {
    yield takeEvery('EDIT_EXPENSE', editExpenseAsync);
}

//WORKERS

export function* addExpenseAsync({ expense, uid }) {
    try {
        const response = yield call(Api.create, expense, uid);  
        yield put({ 
            type: 'ADD_EXPENSE_COMPLETE', 
            expense: { id: response.key, ...expense}
        });      
    }
    catch(error) {
        return;
    }
};

export function* setExpensesAsync(uid) {
    try {
        const expenses = yield call(Api.get, uid);  
        yield put({
            type: 'SET_EXPENSES_COMPLETE',
            expenses
        });      
    }
    catch(error) {
        return;
    }
};

export function* removeExpenseAsync({ id, uid }) {
    try {
        yield call(Api.remove, id, uid);  
        yield put({
            type: 'REMOVE_EXPENSE_COMPLETE',
            id
        });      
    }
    catch(error) {
        return;
    }
};

export function* editExpenseAsync({ updates, id, uid }) {
    try {
        yield call(Api.edit, updates, id, uid); 
        yield put({
            type: 'EDIT_EXPENSE_COMPLETE',
            id,
            updates
        });       
    }
    catch(error) {
        return;
    }
};