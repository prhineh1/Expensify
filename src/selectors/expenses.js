import moment from 'moment';
import partition from 'js-partition';

//Get visible expenses

export default (expenses, {text, sortBy, startDate, endDate, pagination, expensesPartitionIndex}) => {
    let partitionedExpenses = expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return b.createdAt - a.createdAt;
        } else if (sortBy === 'amount') {
            return b.amount - a.amount;
        }
    });
    
    if (partitionedExpenses.length % pagination === 0 && partitionedExpenses.length > 0) {
        partitionedExpenses = partition(pagination, pagination, partitionedExpenses);
    }
    else {
        partitionedExpenses = partition(pagination, pagination, [], partitionedExpenses);        
    }

    //first return value is a jagged array with all the expenses
    //second return value is the current array that will be displayed
    return [partitionedExpenses, partitionedExpenses[expensesPartitionIndex]];
};