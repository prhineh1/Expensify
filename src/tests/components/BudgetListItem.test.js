import React from 'react';
import { shallow } from 'enzyme';
import BudgetListItem from '../../components/BudgetListItem';
import expenses from '../fixtures/expenses';
import budget from '../fixtures/budgets';

test('should render BudgetListItem with an expense', () => {
  const wrapper = shallow(<BudgetListItem budget={budget} expenses={[expenses[1]]} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render BudgetListItem without an expense', () => {
  const wrapper = shallow(<BudgetListItem budget={budget} expenses={[]} />);
  expect(wrapper).toMatchSnapshot();
});