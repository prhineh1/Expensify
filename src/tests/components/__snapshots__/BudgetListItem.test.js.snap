// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render BudgetListItem with an expense 1`] = `
<Link
  className="list-item"
  replace={false}
  to="/edit/budget/a;sdlfj2089"
>
  <div>
    <h3
      className="list-item__title"
    >
      utilities
    </h3>
  </div>
  <h4
    className="list-item__budget"
  >
    <span
      className="list-item__budget__mobile"
    >
      Expenses: 
      $100.95
    </span>
    <span
      className="list-item__budget__desktop"
    >
      $100.95
    </span>
  </h4>
  <h3
    className="list-item__data"
  >
    <span
      className="list-item__data__mobile"
    >
      Budget: 
      $200.00
    </span>
    <span
      className="list-item__data__desktop"
    >
      $200.00
    </span>
  </h3>
</Link>
`;

exports[`should render BudgetListItem without an expense 1`] = `
<Link
  className="list-item"
  replace={false}
  to="/edit/budget/a;sdlfj2089"
>
  <div>
    <h3
      className="list-item__title"
    >
      utilities
    </h3>
  </div>
  <h4
    className="list-item__budget"
  >
    <span
      className="list-item__budget__mobile"
    >
      Expenses: 
      $0.00
    </span>
    <span
      className="list-item__budget__desktop"
    >
      $0.00
    </span>
  </h4>
  <h3
    className="list-item__data"
  >
    <span
      className="list-item__data__mobile"
    >
      Budget: 
      $200.00
    </span>
    <span
      className="list-item__data__desktop"
    >
      $200.00
    </span>
  </h3>
</Link>
`;
